### PR TITLE
Feat/process ocr results

### PR DESCRIPTION
**Pull request by eneemann**
This PR adds two new functions and CLI options.  The first one (`download_ocr_results`) downloads and combines the zipped parquet files into a single zipped parquet file using pandas and dataframe concatenation.  It also deduplicates rows and deletes the original files, leaving only the combined file in local storage.  The second one (`filter_ocr_results`) does some initial filtering of the data to remove extra spaces, newlines, and special characters and save the data locally.  Entries that are completely alphabetical or start with a letter are removed.  It also calculates the original UDOT filename in a new column (ex: 'archive1/001-4_01_Plan_Rev-1.TIF').  This isn't a finished product, but the initial results seem pretty good.

- Some numbers (not sure why I seem to have more OCR rows than valid mosaics):
  - Total files: 89,537
  - Number of valid mosaic files: 52,106
  - Number of rows in combined OCR results: 52,184
  - Number of rows in initial filtered OCR results: 39,938

### CODE REVIEW

**Comment by eneemann** [comment]
Also, I was putting together a function for the CLI to download the spreadsheets (with GUID and project info) and combine them into a single one that could be joined to our results, but ran into some issues.  The spreadsheets don't seem to have the same schemas, so I may need to tweak them manually before they can be combined/joined to the results.

File: row.py [hunk 1]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+        out_dir (str): where to save the results
+
+    Returns:
+        str: the location of the files
+    """
+    if bucket_name.startswith("gs://"):
+        bucket_name = bucket_name[5:]
+
+    if "STORAGE_CLIENT" not in globals():
+        STORAGE_CLIENT = google.cloud.storage.Client()
```
Comment (steveoh) [inline row.py:814]:
```suggestion
```

**todo**: remove this as it circumvents the reasons when the client is created. If you need the client locally, then satisfy the environment variables which allow it to be created.

File: row.py [hunk 2]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+    bucket = STORAGE_CLIENT.bucket(bucket_name)
+    blobs = bucket.list_blobs(prefix=run_name)
+    location = Path(__file__).parent / "data"
+
+    if not location.joinpath(f"ocr_results/{run_name}").exists():
+        location.joinpath(f"ocr_results/{run_name}").mkdir(parents=True)
+
+    ocr_dir = location.joinpath(f"ocr_results/{run_name}")
```
Comment (steveoh) [inline row.py:822]:
```suggestion
    ocr_dir = location.joinpath(f"ocr_results/{run_name}")
    ocr_dir.mkdir(parents=True, exist_ok=True)
```

**suggestion**: it's silly to create the same path twice before putting it in a variable. The if check is unnecessary since mkdir can run with exists_ok whether the directories exist or not.

File: row.py [hunk 3]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+
+    #: download .gz files
+    logging.info("downloading .gz files from cloud storage")
+    for blob in blobs:
+        if blob.name.endswith(".gz"):
+            print(blob.name)
```
Comment (steveoh) [inline row.py:828]: **todo**: remove or replace print statements

File: row.py [hunk 4]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+            blob.download_to_filename(ocr_dir / blob.name)
```
Comment (steveoh) [inline row.py:829]: **polish**: the list blobs iterator should only contain `.gz` files. If that is not the case then the filter should be updated. 

This could also be a comprehension. Ideally we update this so the if filter isn't necessary.

```suggestion
    [blob.download_to_filename(ocr_dir / blob.name) for blob in blobs if blob.name.endswith(".gz")]
```

File: row.py [hunk 5]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+
+    iterator = ocr_dir.glob("*.gz")
+
+    ocr_files = [item for item in iterator if item.is_file()]
```
Comment (steveoh) [inline row.py:833]: **question**: why is this necessary? Shouldn't all the files in the ocr_dir be *.gz files?

File: row.py [hunk 6]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+    logging.info("removed %i duplicate rows", diff)
+
+    out_file = Path(out_dir) / "ocr_results" / "combined_ocr_results.gz"
+    combined_df.to_parquet(out_file, compression="gzip")
+    logging.info("saved combined ocr results to %s", out_file)
+
+    #: delete downloaded files
+    logging.info("deleting individual ocr files")
+    for ocr_file in ocr_files:
+        Path(ocr_file).unlink()
```
Comment (steveoh) [inline row.py:860]: **polish**: convert to comprehension

File: row.py [hunk 7]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+
+    return out_dir
+
+
+def filter_ocr_results(original_results_file, out_dir):
+    """download ocr results from a GCP bucket
```
Comment (steveoh) [inline row.py:866]: **todo**: update the doc string

File: row.py [hunk 8]
```diff
@@ -795,6 +796,143 @@ def download_run(bucket, run_name):
...
+    Returns:
+        str: the location of the output CSV file
+    """
+    #: silence pandas SettingWithCopyWarning
+    pd.options.mode.chained_assignment = None
+
+    out_dir = Path(out_dir)
+
+    if not out_dir.exists():
+        out_dir.mkdir(parents=True)
```
Comment (steveoh) [inline row.py:881]: **todo**: replace if with exists_ok

**Review by steveoh — COMMENTED** [review]
This looks great. A few comments to resolve before merging. I assume you've used this output to be pretty close to create the deliverable for UDOT?

File: row.py (hunk 2 above)
Comment (eneemann) [reply row.py]: done

File: row.py (hunk 4 above)
Comment (eneemann) [reply row.py]: done

File: row.py (hunk 6 above)
Comment (eneemann) [reply row.py]: done

File: row.py (hunk 3 above)
Comment (eneemann) [reply row.py]: done

File: row.py (hunk 8 above)
Comment (eneemann) [reply row.py]: done

**Comment by eneemann** [comment]
@steveoh take another look and see what you think

**Review by steveoh — APPROVED** [review]
